tally/src/counter_impl.cc

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <string>
+
+#include "tally/src/counter_impl.h"
+
+namespace tally {
+
+CounterImpl::CounterImpl() : current_(0), previous_(0) {}
+
+void CounterImpl::Inc() { Inc(1); }
+
+void CounterImpl::Inc(int64_t delta) { current_ += delta; }
+
+void CounterImpl::Report(const std::string &name,
+                         const std::map<std::string, std::string> &tags,
+                         StatsReporter *reporter) {
+  auto val = Value();
+  if (val != 0 && reporter != nullptr) {
+    reporter->ReportCounter(name, tags, val);
+  }
+}
+
+int64_t CounterImpl::Value() {
+  // Load the previous value second so we don't lose any increments.
+  auto current = current_.load();
+  auto previous = previous_.load();
+
+  if (current == previous) {
+    return 0;
+  }
+
+  previous_ = current;
```

Comment:
This seems like a race. The transaction here is not guaranteed to be thread-safe at all. So if I call value in one thread and in another thread concurrently, we have no idea which will set the value of `previous_`. A mutex might even be faster than an atomic value here. See this slidedeck presentation from CPPCon 2017: https://github.com/CppCon/CppCon2017/blob/master/Presentations/C%2B%2B%20Atomics%2C%20From%20Basic%20to%20Advanced/C%2B%2B%20Atomics%2C%20From%20Basic%20to%20Advanced%20-%20Fedor%20Pikus%20-%20CppCon%202017.pdf. Has some good graphs showing mutex is better than atomic.

Reply:
That's a great point! I think switching to using the mutex is the right idea, will address.

Reply:
The race isn't triggered because `counter->Report` is only ever called by the parent scope's `Report()` which is already guarded a mutex. Plus, these are internal APIs so a user triggering such a case isn't a concern. All this to say acquiring a lock for each `counter.Value()` is going to add unnecessary overhead. Please revert to 2 the two atomic impl.

Reply:
Imo the implementation should guarantee that the method is race-free not the caller. I think a fair compromise would be to use an atomic for `current_` so we can continue to increment it without needing to acquire a lock but lock `previous_` to ensure concurrent calls to `Value` don't overwrite one another.

Reply:
@prateek look at the PDF. You probably get *better* performance using a mutex instead of atomics. The only reason to use atomics is if you care about non-blocking algorithms where you need to guarantee wait-freedom/lock-freedom, etc. Let's see a benchmark before reverting it.

If you must revert, there is a way to do this atomically. Use `compare_exchange_strong` when assigning `current_ = previous_`. But I'm not convinced about performance.

Reply:
@jeromefroe I don't follow. The implementation does guarantee it - There are no racy-calls to `previous_` in the scope/reporter and no publicly callable methods can trigger the race. Why add the lock on top of that?

Reply:
@isaachier in that PDF, slide 33 - number of operations/s for atomics is higher than for a mutex. What are you talking about?

If you're not convinced, feel free to add a benchmark.

Reply:
@prateek What I was saying is that the object should ensure the method is thread-safe. Previously that wasn't the case, so even if the total implementation is thread safe now it could be very easy to break that in the future. As a compromise, I'll update this to use @isaachier suggestion of using`compare_exchange_strong`.

Reply:
Look at slides 5-6 @prateek. It depends on the situation but it doesn't scale as well as mutexes do.

Reply:
Hmm, on second thought, `compare_exchange_strong` is also subject to the same races as previously. I think our best option is to make `current_` atomic so increments don't need to acquire a lock but protect `previous_` with a lock to ensure `Value` also returns consistent values.

Reply:
`compare_exchange_strong` guarantees that an older transaction does not overwrite a newer transaction, but it won't guarantee 100% safe updates. However, my benchmarks did show in this case atomics are probably better (at least on Mac). It depends on what you need in this case. Thread safety may not be a realistic concern.

Reply:
@jeromefroe I still think we should use the two atomic solution. All other clients - Go/Java/existing C++ - do so, and we should stay uniform as far as possible. Would documenting the requirements on the caller in the `CounterImp::Value()` satisfy your concerns?

Reply:
@isaachier as pointed out earlier, there is no question of scaling - the Value() method is only called by a single thread.

Reply:
@prateek that implementation is racy so I don't think uniformity applies in this case. Documenting the behavior is certainly better than nothing, but why not start with the safe API and revisit later if it becomes a bottleneck?

Reply:
As a compromise I removed the mutex and added comments that `Value` and `Report` should only be called from a single thread. I still feel that's not the right move but I also don't want to get hung up on this.